dask/bytes/core.py

```diff
@@ -1,22 +1,152 @@
 
 import io
+import os
 
 from toolz import merge
+from warnings import warn
 
 from .compression import seekable_files, files as compress_files
 from .utils import SeekableFile
-from ..compatibility import PY2
+from ..compatibility import PY2, unicode
 from ..delayed import delayed
-from ..utils import infer_storage_options, system_encoding
+from ..utils import (infer_storage_options, system_encoding,
+                     build_name_function, infer_compression)
 
 delayed = delayed(pure=True)
 
 # Global registration dictionaries for backend storage functions
 # See docstrings to functions below for more information
 _read_bytes = dict()
+_write_bytes = dict()
 _open_files = dict()
 _open_text_files = dict()
+_write_files = dict()
+
+
+def write_block_to_file(data, f, compression, encoding):
+    """
+    Parameters
+    ----------
+    data : data to write
+        Either str/bytes, or iterable producing those, or something file-like
+        which can be read.
+    f : file-like
+        backend-dependent file-like object
+    compression : string
+        a key of `compress_files`
+    encoding : string (None)
+        if a string (e.g., 'ascii', 'utf8'), implies text mode, otherwise no
+        encoding and binary mode.
+    """
+    original = False
+    f2 = f
+    f = SeekableFile(f)
+    if compression:
+        original = True
+        f = compress_files[compression](f, mode='wb')
+    try:
+        if isinstance(data, (str, bytes)):
+            if encoding:
+                f.write(data.encode(encoding=encoding))
+            else:
+                f.write(data)
+        elif isinstance(data, io.IOBase):
+            # file-like
+            out = '1'
+            while out:
+                out = data.read(64*2**10)
+                if encoding:
+                    f.write(out.encode(encoding=encoding))
+                else:
+                    f.write(out)
+        else:
+            # iterable, e.g., bag contents
+            start = False
+            for d in data:
+                if start:
+                    f.write(b'\n')
+                else:
+                    start = True
+                if encoding:
+                    f.write(d.encode(encoding=encoding))
+                else:
+                    f.write(d)
+    finally:
+        f.close()
+        if original:
+            f2.close()
+
+
+def write_bytes(values, urlpath, name_function=None, compression=None,
+                lazy=False, encoding=None, **kwargs):
+    """For a list of values which evaluate to byte, produce delayed values
+    which, when executed, result in writing to files.
+
+    The path maybe a concrete directory, in which case it is interpreted
+    as a directory, or a template for numbered output.
+
+    The path may be preceded by a protocol, like ``s3://`` or ``hdfs://`` if
+    those libraries are installed.
+
+    Parameters
+    ----------
+    values: list of Values or dask collection
+        the data to be written
```

Comment:
list of *dask.delayed* values or ...

We stopped using the class name `Values`, which has since been renamed `Delayed`